core/src/main/scala/kafka/utils/ZkUtils.scala

```diff
@@ -360,25 +350,14 @@ class ZkUtils(val zkClient: ZkClient,
   def registerBrokerInZk(id: Int,
                          host: String,
                          port: Int,
-                         advertisedEndpoints: collection.Map[SecurityProtocol, EndPoint],
+                         advertisedEndpoints: Seq[EndPoint],
                          jmxPort: Int,
                          rack: Option[String],
                          apiVersion: ApiVersion) {
     val brokerIdPath = BrokerIdsPath + "/" + id
-    val timestamp = Time.SYSTEM.milliseconds.toString
-
-    val version = if (apiVersion >= KAFKA_0_10_0_IV1) 3 else 2
-    var jsonMap = Map("version" -> version,
-                      "host" -> host,
-                      "port" -> port,
-                      "endpoints" -> advertisedEndpoints.values.map(_.connectionString).toArray,
-                      "jmx_port" -> jmxPort,
-                      "timestamp" -> timestamp
-    )
-    rack.foreach(rack => if (version >= 3) jsonMap += ("rack" -> rack))
-
-    val brokerInfo = Json.encode(jsonMap)
-    registerBrokerInZk(brokerIdPath, brokerInfo)
+    val version = if (apiVersion >= KAFKA_0_10_0_IV1) 4 else 2
```

Comment:
It looks weird, but it is actually right, I think. The method documentation explains it:

```java
* Register brokers with v4 json format (which includes multiple endpoints and rack) if
   * the apiVersion is 0.10.0.X or above. Register the broker with v2 json format otherwise.
   * Due to KAFKA-3100, 0.9.0.0 broker and old clients will break if JSON version is above 2.
   * We include v2 to make it possible for the broker to migrate from 0.9.0.0 to 0.10.0.X or above without having to
   * upgrade to 0.9.0.1 first (clients have to be upgraded to 0.9.0.1 in any case).
   *
   * This format also includes default endpoints for compatibility with older clients.
```

We only need to keep the version low for 0.9.0.0 brokers that had a bug in their parsing logic. More recent brokers will happily read newer versions as they expect no incompatible changes  in the format of the existing fields.

Does that make sense?